cpp/src/arrow/compare.cc

```diff
@@ -1071,6 +1065,67 @@ bool TensorEquals(const Tensor& left, const Tensor& right) {
   return are_equal;
 }
 
+template <typename DataType>
+bool StridedFloatTensorContentEquals(int dim_index, int64_t left_offset, int64_t right_offset,
+                                     const Tensor& left, const Tensor& right) {
+  using c_type = typename DataType::c_type;
+  const int64_t n = left.shape()[dim_index];
+  if (dim_index == left.ndim() - 1) {
+    auto left_data = left.raw_data();
+    auto right_data = right.raw_data();
+    auto left_stride = left.strides()[dim_index];
+    auto right_stride = right.strides()[dim_index];
+    for (int64_t i = 0; i < n; ++i) {
+      c_type left_value = *reinterpret_cast<const c_type*>(left_data + left_offset + i * left_stride);
+      c_type right_value = *reinterpret_cast<const c_type*>(right_data + right_offset + i * right_stride);
+      if (left_value != right_value || std::isnan(left_value) || std::isnan(right_value)) {
+        return false;
+      }
+    }
+    return true;
+  }
+  for (int64_t i = 0; i < n; ++i) {
+    if (!StridedFloatTensorContentEquals<DataType>(dim_index + 1, left_offset, right_offset, left, right)) {
+      return false;
+    }
+    left_offset += left.strides()[dim_index];
+    right_offset += right.strides()[dim_index];
+  }
+  return true;
+}
+
+template <typename DataType>
+bool FloatTensorEquals(const Tensor& left, const Tensor& right) {
+  static_assert(std::is_floating_point<typename DataType::c_type>::value,
+                "DataType must be a floating point type");
+  return StridedFloatTensorContentEquals<DataType>(0, 0, 0, left, right);
+}
+
+}
+
+bool TensorEquals(const Tensor& left, const Tensor& right) {
+  if (left.type_id() != right.type_id()) {
+    return false;
+  } else if (left.size() == 0) {
+    return true;
```

Comment:
I guess this isn't an appropriate return value.

Instead, should we change this clause to the following?

```c++
} else if (left.size() == 0 && right.size() == 0) {
  return true;
}
```